[visionOS] REGRESSION(299348@main): Timeline on inline video controls doesn't glow anymore, making it hard to interact with

etiennesegonzac opened the pull request:
#### cab48a00b61664f3704bbad2075487aff0d9883b
<pre>
[visionOS] REGRESSION(299348@main): Timeline on inline video controls doesn&apos;t glow anymore, making it hard to interact with
<a href="https://bugs.webkit.org/show_bug.cgi?id=301196">https://bugs.webkit.org/show_bug.cgi?id=301196</a>
&lt;<a href="https://rdar.apple.com/161134512">rdar://161134512</a>&gt;

Reviewed by Jer Noble.

Go back to setting the TimeControl&apos;s children at layout time in order
for the VisionSlider customization to be picked up.

Add support for flexbox layouts to the VisionSlider.

Test: media/modern-media-controls/visionos-inline-media-controls/visionos-inline-media-controls-layout.html

* LayoutTests/TestExpectations:
* LayoutTests/media/modern-media-controls/visionos-inline-media-controls/visionos-inline-media-controls-layout-expected.txt: Added.
* LayoutTests/media/modern-media-controls/visionos-inline-media-controls/visionos-inline-media-controls-layout.html: Added.
* LayoutTests/platform/visionos/TestExpectations:
Add a basic visionOS specific test to cover the regression.

* Source/WebCore/Modules/modern-media-controls/controls/time-control.js:
(TimeControl.):
Don&apos;t set the children from the constructor.
(TimeControl.prototype._performIdealLayout):
Set children during layout like we did before 299348@main.

(TimeControl.prototype.this.remainingTimeLabel): Deleted.
* Source/WebCore/Modules/modern-media-controls/controls/vision-slider.js:
(VisionSlider.prototype.commit):
When part of ta flexbox layout, the width is not set on the Slider.
Add support for this setup using a percentage based width for the
primary fill.

* Source/WebCore/Modules/modern-media-controls/js-files:
Make the `VisionMediaControls` testable.

Canonical link: <a href="https://commits.webkit.org/302252@main">https://commits.webkit.org/302252@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a7fb7b93abc8864165bcf68e40c267db0dea058

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128476 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/744 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39307 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135869 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79919 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/689 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/618 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97804 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65717 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/89a25d49-eff6-405e-8853-11f2e0a5fbf9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131424 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/505 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115107 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78415 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b760787f-9336-487d-86d2-5b4ae8acb310) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/456 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33215 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79152 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108872 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33698 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138318 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/582 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/550 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106342 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/626 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111448 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106153 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27048 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/486 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29988 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52922 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/638 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63834 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/528 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/590 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/594 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/Modules/modern-media-controls/controls/vision-slider.js:
```diff
@@ -68,11 +68,16 @@ class VisionSlider extends SliderBase
+            const fillWidth = this.value * (this.width - (margin * 2) - height) + height;
+            this._primaryFill.element.style.width = `${fillWidth}px`;
+        } else
+            this._primaryFill.element.style.width = `calc((100% - ${height}px) * ${this.value} + ${height}px)`;
```
Nit: surely there's a better way to implement this than explicit pixel values? We should revisit most of these places where layout depends on calculating pixel values; this causes general slowdowns in layout. Especially for the "could be hosted in a flex context", we should be able to add static CSS that lays this out correctly.

etiennesegonzac replied:
Made it clearer in the code that we're just using the css 100% as replacement for `this.width` in the formula, and filed a follow-up to use flexGrow in both cases.

jernoble reviewed (approved):
r=me with nit

etiennesegonzac committed:
[visionOS] REGRESSION(299348@main): Timeline on inline video controls doesn't glow anymore, making it hard to interact with
https://bugs.webkit.org/show_bug.cgi?id=301196
<rdar://161134512>

Reviewed by Jer Noble.

Go back to setting the TimeControl's children at layout time in order
for the VisionSlider customization to be picked up.

Add support for flexbox layouts to the VisionSlider.

Test: media/modern-media-controls/visionos-inline-media-controls/visionos-inline-media-controls-layout.html

* LayoutTests/TestExpectations:
* LayoutTests/media/modern-media-controls/visionos-inline-media-controls/visionos-inline-media-controls-layout-expected.txt: Added.
* LayoutTests/media/modern-media-controls/visionos-inline-media-controls/visionos-inline-media-controls-layout.html: Added.
* LayoutTests/platform/visionos/TestExpectations:
Add a basic visionOS specific test to cover the regression.

* Source/WebCore/Modules/modern-media-controls/controls/time-control.js:
(TimeControl.):
Don't set the children from the constructor.
(TimeControl.prototype._performIdealLayout):
Set children during layout like we did before 299348@main.

(TimeControl.prototype.this.remainingTimeLabel): Deleted.
* Source/WebCore/Modules/modern-media-controls/controls/vision-slider.js:
(VisionSlider.prototype.commit):
When part of ta flexbox layout, the width is not set on the Slider.
Add support for this setup using a percentage based width for the
primary fill.

* Source/WebCore/Modules/modern-media-controls/js-files:
Make the `VisionMediaControls` testable.

Canonical link: https://commits.webkit.org/302252@main
